### PR TITLE
WIP Test/Rust integration for arbitrary logic

### PR DESCRIPTION
**Pull request by jkilpatr**
TODO:

- [x] deploy arbitrary logic contract in tests
- [x] hardhat in test environment
- [x] New endblocker for the Cosmos module to simulate arbitrary logic user
- [x] getters/setters and parsing for go endpoints
- [x] env var passing for alchemy ID
- [x] test environment for arbitrary logic
- [x] encoding for the logic call confirm message
- [x] argument encoding for the logic call endpoint
- [x] logic to relay logic batch executed event
- [x] logic to sign confirms
- [x] logic to relay logic calls
- [ ] slashing for arbitrary logic calls
- [x] genesis state save/load for arbitrary logic call state
- [ ] testing of messages / logic

### CODE REVIEW

File: orchestrator/cosmos_peggy/src/send.rs [hunk 1]
```diff
@@ -166,12 +166,58 @@ pub async fn send_batch_confirm(
     contact.retry_on_block(tx).await
 }
 
+/// Send in a confirmation for a specific logic call
```
Comment (zmanian) [inline orchestrator/cosmos_peggy/src/send.rs:169]:
```suggestion
/// Send a ethereum signature to the Cosmos that confirms the logic call and provides the required material to relay.
```

File: orchestrator/ethereum_peggy/src/message_signatures.rs [hunk 2]
```diff
@@ -160,3 +158,84 @@ fn test_batch_signature() {
...
+    let mut fee_token_contracts = Vec::new();
+    for item in call.transfers.iter() {
+        transfer_amounts.push(Token::Uint(item.amount.clone()));
+        transfer_token_contracts.push(item.token_contract_address);
+    }
+    for item in call.fees.iter() {
+        fee_amounts.push(Token::Uint(item.amount.clone()));
+        fee_token_contracts.push(item.token_contract_address);
+    }
+
```
Comment (zmanian) [inline orchestrator/ethereum_peggy/src/message_signatures.rs:180]:
```suggestion
// We are encoding for the following ABI.
//            { "internalType": "uint256[]", "name": "transferAmounts",        "type": "uint256[]" },
//            { "internalType": "address[]", "name": "transferTokenContracts", "type": "address[]" },
//            { "internalType": "uint256[]", "name": "feeAmounts",             "type": "uint256[]" },
//            { "internalType": "address[]", "name": "feeTokenContracts",      "type": "address[]" },
//            { "internalType": "address",   "name": "logicContractAddress",   "type": "address"   },
//            { "internalType": "bytes",     "name": "payload",                "type": "bytes"     },
//            { "internalType": "uint256",   "name": "timeOut",                "type": "uint256"   },
//            { "internalType": "bytes32",   "name": "invalidationId",         "type": "bytes32"   },
//            { "internalType": "uint256",   "name": "invalidationNonce",      "type": "uint256"   }
```

File: orchestrator/ethereum_peggy/src/message_signatures.rs [hunk 3]
```diff
@@ -160,3 +158,84 @@ fn test_batch_signature() {
...
+        Token::FixedString("logicCall".to_string()),
+        Token::Dynamic(transfer_amounts),
+        transfer_token_contracts.into(),
+        Token::Dynamic(fee_amounts),
+        fee_token_contracts.into(),
+        call.logic_contract_address.into(),
+        Token::UnboundedBytes(call.payload),
+        call.timeout.into(),
+        Token::Bytes(call.invalidation_id),
+        call.invalidation_nonce.into(),
```
Comment (zmanian) [inline orchestrator/ethereum_peggy/src/message_signatures.rs:192]:
```suggestion
        Token::FixedString(peggy_id), // Peggy Instance ID
        Token::FixedString("logicCall".to_string()), //Function Name
        Token::Dynamic(transfer_amounts), //Array of Transfer amounts
        transfer_token_contracts.into(), //ERC-20 contract for transfers
        Token::Dynamic(fee_amounts), // Array of Fees
        fee_token_contracts.into(), // ERC-20 contract for fee payments
        call.logic_contract_address.into(), // Address of a logic contract
        Token::UnboundedBytes(call.payload), // Encoded arguments to logic contract
        call.timeout.into(), // Timeout on batch
        Token::Bytes(call.invalidation_id), // ID of logic batch
        call.invalidation_nonce.into(), // Nonce of logic batch. See 2-d nonce scheme.
```

**Comment by zmanian** [comment]
utACK

File: orchestrator/ethereum_peggy/src/message_signatures.rs (hunk 3 above)
Comment (jkilpatr) [reply orchestrator/ethereum_peggy/src/message_signatures.rs:192]: good call, will apply :+1:

**Comment by jkilpatr** [comment]
this has been merged. The test suite is not quite complete as I couldn't find the time to actually write a working execution example. The challenge here is the tradeoff between wanting a minimal test to keep within the gravity codebase and actually properly exercising the complexity of an arbitrary logic call.

It's the job of the implementing module to figure out what contract it's calling, what data it needs to encode and all of these details. I don't want the test to become a full fledged user implementation of this feature. I think the best bet is to have hardhat deploy an erc20 always at the same address then make some simple call like 'approve' on it.

**Pull request closed without merging**